PixelGameEngine Tutorials

GaxysGalactic opened the pull request:
Includes all the PixelGameEngine tutorials done so far:
- Flappy Bird
- Asteroids
- Pathfinding (A*)

GaxysGalactic committed:
FlappyBird Progress Day 1

It kinda works. Collision detection needs to be better and better to rename stuff and store stuff better

GaxysGalactic committed:
Working Game!

Need to refactor this ASAP.

GaxysGalactic committed:
OOP and Organization

+ Added objects to better represent some variable groups
o Moved some stuff into their own functions

GaxysGalactic committed:
Asteroids Part 1

Part 1 of Asteroids Game, missing about half of it

GaxysGalactic committed:
Refactoring Asteroids

o Condensed a lot of different functionalities into their own functions

GaxysGalactic committed:
Pathfinding Complete!

+ Added Pathfinding project
o Refactored previous code to better match coding standards

GaxysGalactic committed:
Final fixes

+ Added Section lines
+ Added better comments
o Changes to some methods

DvdHalliday commented on Projects/PixelGameEngine/PixelTutorials/Pathfinding/main.cpp:
```diff
@@ -0,0 +1,353 @@
+public:
+	Pathfinding()
+	{
+		// Name your application
```
Unnecessary comment

DvdHalliday commented on Projects/PixelGameEngine/PixelTutorials/Pathfinding/main.cpp:
```diff
@@ -0,0 +1,353 @@
+		int x;
+		int y;
+		std::vector<FNode*> Neighbours;
+		FNode* Parent;
```
Maybe add some separating blank lines to group common attributes?

DvdHalliday commented on Projects/PixelGameEngine/PixelTutorials/Pathfinding/main.cpp:
```diff
@@ -0,0 +1,353 @@
+		float LocalGoal;
+		int x;
+		int y;
+		std::vector<FNode*> Neighbours;
```
our coding standards state we should be using US spelling for variable names

DvdHalliday commented on Projects/PixelGameEngine/PixelTutorials/Pathfinding/main.cpp:
```diff
@@ -0,0 +1,353 @@
+		bool bVisited = false;
+		float GlobalGoal;
+		float LocalGoal;
+		int x;
+		int y;
```
I'm not sure but... according to standards maybe these two should be capitilised?

simonkytr replied:
or maybe you should name like "PositionX"

DvdHalliday commented on Projects/PixelGameEngine/PixelTutorials/Pathfinding/main.cpp:
```diff
@@ -0,0 +1,353 @@
+
+	const int NodeSize = 9;
+	const int NodeBorder = 2;
+	int NodeInner = NodeSize - (NodeBorder * 2);
```
>;o

DvdHalliday commented on Projects/PixelGameEngine/PixelTutorials/Pathfinding/main.cpp:
```diff
@@ -0,0 +1,353 @@
+	const int MapHeight = 16;
+
+	FNode* NodeStart = nullptr;
+	FNode* NodeEnd = nullptr;
```
Perhaps Node at the end of the variable name would make more sense? i think Javid names them like these 'cause of his own naming conventions, just as he names vecNodes... and such, bur for us I think StartNode would probably be best, what do you think?

DvdHalliday commented on Projects/PixelGameEngine/PixelTutorials/Pathfinding/main.cpp:
```diff
@@ -0,0 +1,353 @@
+	}
+
+	//-------------------------------------------------------------------------------------------
+	void ConnectOrthogonally(const int& x, const int& y)
```
built-in parameters shouldn't be referenced

DvdHalliday commented on Projects/PixelGameEngine/PixelTutorials/Pathfinding/main.cpp:
```diff
@@ -0,0 +1,353 @@
+	}
+
+	//-------------------------------------------------------------------------------------------
+	void ConnectDiagonally(const int& x, const int& y)
```
built-in parameters shouldn't be referenced

DvdHalliday commented on Projects/PixelGameEngine/PixelTutorials/Pathfinding/main.cpp:
```diff
@@ -0,0 +1,353 @@
+		{
+			for (int y = 0; y < MapHeight; y++)
+			{
+				//Either of these can be commented out. Commenting out both technically also works.
+
+				ConnectOrthogonally(x, y);
+
+				ConnectDiagonally(x, y);
+				
```
Interesting, however if you leave both commented out you're gonna' create double connections between each node, won't you?

GaxysGalactic replied:
If I leave both commented, there are no connections between nodes. It looks kinda sad. 

![image](https://user-images.githubusercontent.com/44332093/194943663-bdf7ce0d-1c64-46aa-8e81-178f9722183c.png)

sancho6040 replied:
Why would you comment both methods if this will cause the program to do nothing?? Xd

DvdHalliday commented on Projects/PixelGameEngine/PixelTutorials/Pathfinding/main.cpp:
```diff
@@ -0,0 +1,353 @@
+	//-------------------------------------------------------------------------------------------
+	void UpdateGridStates()
+	{
+		//Use integer division to nicely get cursor position in node space
```
BUT HOW NICELY IS IT REALLY, HUH?

DvdHalliday commented on Projects/PixelGameEngine/PixelTutorials/Pathfinding/main.cpp:
```diff
@@ -0,0 +1,353 @@
+		//Reset Navigation Graph - default all node states
+		ResetNavGraph();
+
+		//Lambda Functions, for convenience
+		auto Distance = [](FNode* a, FNode* b) 
+		{
+			return sqrtf((a->x - b->x) * (a->x - b->x) + (a->y - b->y) * (a->y - b->y));
+		};
```
Lovely

DvdHalliday commented on Projects/PixelGameEngine/PixelTutorials/Pathfinding/main.cpp:
```diff
@@ -0,0 +1,353 @@
+					//The best path length to the neighbour being tested has changed, so update the neighbour's score.
+					//The heuristic is used to globally bias the path algorithm, so it knows if its getting better or worse.
+					//At some point, the algorithm will realize this path is worse and abandon it, and then go and search along the next best path.
+					Neighbour->GlobalGoal = Neighbour->LocalGoal + Heuristic(Neighbour, NodeEnd);
```
looks good to me

camodesigns replied:
for me too, so cool

DvdHalliday commented on Projects/PixelGameEngine/PixelTutorials/Pathfinding/main.cpp:
```diff
@@ -0,0 +1,353 @@
+	}
+
+	//-------------------------------------------------------------------------------------------
+	void SolveAStar()
```
It looks good, however yup I think it's a bit long so you might want to consider separating some chunks of it into different functions

DvdHalliday reviewed (changes requested):
Good job!

This is some clean code right here. There's a couple of comments you might wanna check

simonkytr commented on Projects/PixelGameEngine/PixelTutorials/Pathfinding/main.cpp:
```diff
@@ -0,0 +1,353 @@
+#include "olcPixelGameEngine.h"
+
+// Override base class with your custom functionality
+class Pathfinding : public olc::PixelGameEngine
```
according to the coding standards, you have to name this class with the f prefix

camodesigns replied:
F in Capital Letters

camodesigns commented on Projects/PixelGameEngine/PixelTutorials/Pathfinding/main.cpp:
```diff
@@ -0,0 +1,353 @@
+	}
+
+	//-------------------------------------------------------------------------------------------
+	bool OnUserUpdate(float ElapsedTime) override
```
The name could be more specific, javi does it that way but it is not what we are looking for

camodesigns commented on Projects/PixelGameEngine/PixelTutorials/Pathfinding/main.cpp:
```diff
@@ -0,0 +1,353 @@
+public:
+
+	//-------------------------------------------------------------------------------------------
+	bool OnUserCreate() override
```
The name could be more specific, javi does it that way but it is not what we are looking for

GaxysGalactic committed:
Code Review fixes and Pathfinding 2

+ Added Pathfinding 2. This has not been refactored, due to moving on to the project.
o Fixed Pathfinding 1 according to code review

GaxysGalactic committed:
More fixes

sancho6040 commented on Projects/PixelGameEngine/PixelTutorials/Pathfinding/main.cpp:
```diff
@@ -0,0 +1,356 @@
+	//-------------------------------------------------------------------------------------------
+	bool OnUserUpdate(float ElapsedTime) override
+	{
+		//Update the grid according to mouse input
+		UpdateGridStates();
+
+		//Clear Screen
+		FillRect(0, 0, ScreenWidth(), ScreenHeight(), olc::BLACK);
+
+		//Draw Connections First - lines from this nodes position to its connected neighbour node positions
+		DrawConnections();
+
+		//Draw Nodes on top
+		DrawNodes();
+
+		//Draw Path by starting at the end, and following the parent node trail back to the start - the start node will not have a parent path to follow
+		DrawPath();
+
+		return true;
```
great work with the organization of your code 👌

sancho6040 commented on Projects/PixelGameEngine/PixelTutorials/Pathfinding/main.cpp:
```diff
@@ -0,0 +1,356 @@
+		//Reset Navigation Graph - default all node states
+		ResetNavGraph();
+
+		//Lambda Functions, for convenience
+		auto Distance = [](FNode* a, FNode* b) 
+		{
+			return sqrtf((a->x - b->x) * (a->x - b->x) + (a->y - b->y) * (a->y - b->y));
+		};
```
May be here you can use the method pow from "Math. h", this will make the code more readable.

sancho6040 commented on Projects/PixelGameEngine/PixelTutorials/Pathfinding/main.cpp:
```diff
@@ -0,0 +1,356 @@
+			return sqrtf((a->x - b->x) * (a->x - b->x) + (a->y - b->y) * (a->y - b->y));
+		};
+
+		auto Heuristic = [Distance](FNode* a, FNode* b)
```
I don't know about these parameter names, should it be Uppercase? The same for the cases where you use x and y?
What does @JD say about this?

sancho6040 reviewed (commented):
my congratulations to the chef. 👌

JDAlzate commented on Projects/PixelGameEngine/PixelTutorials/Asteroids/main.cpp:
```diff
@@ -0,0 +1,363 @@
+
+	struct SpaceObject
+	{
+		float X;
```
I have no clue what X, Y, dX, and dY is for a SpaceObject 👀. Also dX and dY do not follow conventions.

JDAlzate commented on Projects/PixelGameEngine/PixelTutorials/Asteroids/main.cpp:
```diff
@@ -0,0 +1,363 @@
+
+private:
+
+	struct SpaceObject
```
structs and classes should start with F

JDAlzate commented on Projects/PixelGameEngine/PixelTutorials/Asteroids/main.cpp:
```diff
@@ -0,0 +1,363 @@
+		for (int i = 0; i < Verts; i++)
+		{
+			float Radius = (float)std::rand() / (float)RAND_MAX * 0.4f + 0.8f;
+			float a = ((float)i / (float)Verts) * 6.28318f;
```
What is a 👀

JDAlzate commented on Projects/PixelGameEngine/PixelTutorials/Asteroids/main.cpp:
```diff
@@ -0,0 +1,363 @@
+		}
+	}
+
+	void UpdateShipPhysics(float& ElapsedTime)
```
Why are these floats passed-in as references?

JDAlzate commented on Projects/PixelGameEngine/PixelTutorials/Asteroids/main.cpp:
```diff
@@ -0,0 +1,363 @@
+		WrapCoordinates(Player.X, Player.Y, Player.X, Player.Y);
+
+		//Check ship collision with asteroids
+		for (SpaceObject& a : Asteroids)
```
I don't know what a means. But I can see that it is not being changed, so this should be a const reference 👀 Remember that non-const references are for objects that are intended to be changed.

JDAlzate replied:
I see a trend in the whole project to use unnecessary references for built-in data types and non-const references for instances that don't need to change.

JDAlzate commented on Projects/PixelGameEngine/PixelTutorials/Pathfinding/main.cpp:
```diff
@@ -0,0 +1,356 @@
+	bool OnUserCreate() override
+	{
+		//Dynamic Memory
+		Nodes = new FNode[MapWidth * MapHeight];
```
Smells like a memory leak 👀 I couldn't find a delete anywhere.

Keep in mind that we want to avoid raw pointers and naked new allocations when possible. This is why containers exist, so this memory leak could have been avoided if Nodes wasn't a pointer but rather a vector or some other container.

JDAlzate commented on Projects/PixelGameEngine/PixelTutorials/Pathfinding/main.cpp:
```diff
@@ -0,0 +1,356 @@
+	//-------------------------------------------------------------------------------------------
+	void InitializeNodes()
+	{
+		for (int x = 0; x < MapWidth; x++)
```
I get it may be confusing, but the only acceptable exception for indices are i,j,k since they are universally known. If you're changing the name, you need to follow standards, so this should be an uppercase X, same as the Y.

JDAlzate commented on Projects/PixelGameEngine/PixelTutorials/Pathfinding/main.cpp:
```diff
@@ -0,0 +1,356 @@
+	{
+		if (y > 0)
+		{
+			Nodes[y * MapWidth + x].Neighbors.push_back(&Nodes[(y - 1) * MapWidth + (x + 0)]);
```
This is a good place for helper functions. You're using the `Nodes[y * MapWidth + x]` a lot, and it's not very easy to read. A `GetNode(x, y)` would do wonders for readability.

JDAlzate commented on Projects/PixelGameEngine/PixelTutorials/Pathfinding/main.cpp:
```diff
@@ -0,0 +1,356 @@
+		CreateConnections();
+
+		//Manually positon the start and end markers so they are not null
+		StartNode = &Nodes[(MapHeight / 2) * MapWidth + 1];
```
It's annoying to deal with instances directly and need to use the & operator every time. You could avoid this by having an array of pointers instead of an array of instances, with each node allocated in the heap.

JDAlzate commented on Projects/PixelGameEngine/PixelTutorials/Pathfinding/main.cpp:
```diff
@@ -0,0 +1,356 @@
+		ResetNavGraph();
+
+		//Lambda Functions, for convenience
+		auto Distance = [](FNode* a, FNode* b) 
```
a and b do not follow conventions.

JDAlzate reviewed (commented):
Nice code, but be careful with conventions and calling your variables `n`, `a`, `b`, `x`, etc.
